src/io/flutter/run/test/TestConfigProducer.java

```diff
@@ -35,31 +38,113 @@ protected TestConfigProducer() {
    */
   @Override
   protected boolean setupConfigurationFromContext(TestConfig config, ConfigurationContext context, Ref<PsiElement> sourceElement) {
+    final PsiElement elt = context.getPsiLocation();
+    if (elt instanceof PsiDirectory) {
+      return setupForDirectory(config, (PsiDirectory)elt);
+    }
+
     final DartFile file = FlutterRunConfigurationProducer.getDartFile(context);
-    if (file != null) {
-      return setupForDartFile(config, context, file);
+    if (file == null) {
+      return false;
     }
 
-    final PsiElement elt = context.getPsiLocation();
-    return elt instanceof PsiDirectory && setupForDirectory(config, (PsiDirectory)elt);
+    final String testName = findTestName(elt);
+    if (testName != null) {
+      return setupForSingleTest(config, context, file, testName);
+    }
+
+    return setupForDartFile(config, context, file);
+  }
+
+  /**
+   * Returns the name of the test containing this element, or null if it can't be calculated.
+   */
+  private String findTestName(@Nullable PsiElement elt) {
+    while (elt != null) {
+      // Look for a call expression: test("something", ...);
+      if (elt instanceof DartCallExpression) {
+        final DartCallExpression call = (DartCallExpression) elt;
+        final String name = getCalledFunctionName(call);
+        if (name != null && name.equals("test")) {
+          final DartStringLiteralExpression lit = getFirstArgumentIfStringLiteral(call);
+          if (lit == null) return null;
+          return getStringIfNotTemplate(lit);
+        }
+      }
+      elt = elt.getParent();
+    }
+    return null; // not found
+  }
+
+  @Nullable
+  private String getCalledFunctionName(@NotNull DartCallExpression call) {
+    if (!(call.getFirstChild() instanceof DartReference)) return null;
+    return call.getFirstChild().getText();
```

Comment:
Do we need to consider malformed ASTs?  That is,  might there be a case where `getFirstChild()` returns null?